bin/manage.py

```diff
@@ -0,0 +1,406 @@
+        except Exception as e:
+            # just bail on unexpected exceptions when trying to connect
+            log.error(e)
+            return False
+    return True
+
+@debug
+def mongo_update_replset_config(local_mongo, hostname):
+    '''
+    called from the primary node to update the replica config in mongo
+    using the current set of healthy mongo containers listed in consul
+    '''
+    try:
+        repl_config = local_mongo.admin.command('replSetGetConfig')
+        if not repl_config['ok']:
+            raise Exception('could not get replica config: %s' % repl_config['errmsg'])
+        repl_config = repl_config['config']
+        # TODO use consul.agent.health() instead?
+        consul_services = consul.agent.services()
+
+        mongos_in_consul = []
+        for x in consul_services:
+            host = consul_to_mongo_hostname(consul_services[x]['ID'])
+            if host:
+                mongos_in_consul.append(host)
+        if not len(mongos_in_consul):
+            return
+        if not hostname + ':27017' in mongos_in_consul:
+            return
+
+        max_id = 0
+        changed = False
+        new_members = []
+        for member in repl_config['members']:
+            if member['_id'] > max_id:
+                max_id = member['_id']
+            if member['host'] in mongos_in_consul:
+                mongos_in_consul.remove(member['host'])
+                new_members.append(member)
+            else:
+                changed = True
+        for host in mongos_in_consul:
+            changed = True
+            max_id += 1
+            new_members.append({
+                    '_id': max_id,
+                    'host': host,
+                })
+        
+        if changed:
+            repl_config['members'] = new_members
+            repl_config['version'] += 1
+            local_mongo.admin.command('replSetReconfig', repl_config)
+            log.info('updating replica config in mongo from consul info')
+
+    except Exception as e:
+        log.exception(e)
+        sys.exit(1)
+
+def consul_to_mongo_hostname(name):
+    if name.startswith(PRIMARY + '-'):
+        prefix = PRIMARY + '-'
+    elif name.startswith(SECONDARY + '-'):
+        prefix = SECONDARY + '-'
+    else:
+        return None
+
+    name = name[len(prefix):] + ':27017'
+    return name
+
+# ---------------------------------------------------------
+
+@debug
+def get_primary_node_from_consul(timeout=10):
+    while timeout > 0:
+        try:
+            result = consul.kv.get(PRIMARY_KEY)
+            if result[1]:
+                if result[1].get('Session', False):
+                    return result[1]['Value']
+            # either there is no primary or the session has expired
+            return None
+        except Exception as e:
+            timeout = timeout - 1
+            time.sleep(1)
+    raise e
+
+@debug
+def mark_as_primary(hostname):
+    """ Write flag to Consul to mark this node as primary """
```

Comment:
I'm a little concerned about how we're doing this, because it has a nasty failure mode. In autopilotpattern/mysql, we have this session lock because MySQL has no leader election of its own and we also have an expensive operation to snapshot the data and sync it. Mongo has its own leader election strategy so the only thing this session is doing for us it making sure we initialize the replSet once-and-only once.

But this session doesn't quite ensure we initialize the replSet once-and-only once -- it ensures we initialize the replSet once per primary. If a primary dies then after the session expires the new primary will lock the session on its next health check window. This makes for a window of up to 5 seconds where if a new instance comes up it'll try to initialize a new replica set. What we're really looking for here is the behavior that autopilotpattern/mysql has around the `preStart` hook -- if there's a database snapshot on Manta we get it rather than initializing the local DB. For Mongo I'd suggest something like:

- during `preStart` check for the existence of a snapshot and pull it down if it exists
- during `health` check, if there's no replSet already and there's no primary lock, attempt to obtain a session lock
  - if we get the session (we're primary), initialize the replSet, mongodump the DB to Manta, and write a non-expiring key to Consul that says this has been done
  - if we don't get the session (we're replica), continue with the health check normally